PWGHF/D2H/Macros/ML/train_d2h.py

```diff
@@ -0,0 +1,544 @@
+#!/usr/bin/env python3
+
+# Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+# See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+# All rights not expressly granted are reserved.
+
+# This software is distributed under the terms of the GNU General Public
+# License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+
+# In applying this license CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+
+"""
+file: train_d2h.py
+brief: script for the training of ML models to be used in D2H
+note: inspired by EventFiltering/PWGHF/Macros/train_hf_triggers.py and Run2 macros
+usage: python3 train_d2h.py CONFIG
+author: Alexandre Bigot <alexandre.bigot@cern.ch>, Strasbourg University
+author: Mingyu Zhang <mingyu.zang@cern.ch>, Central China Normal University
+"""
+
+import argparse
+import os
+import pickle
+import sys
+
+import matplotlib.pyplot as plt  # pylint: disable=import-error
+import numpy as np  # pylint: disable=import-error
+import pandas as pd  # pylint: disable=import-error
+import xgboost as xgb  # pylint: disable=import-error
+import yaml  # pylint: disable=import-error
+
+# pylint: disable=import-error
+from hipe4ml import plot_utils
+from hipe4ml.model_handler import ModelHandler
+from hipe4ml.tree_handler import TreeHandler
+from hipe4ml_converter.h4ml_converter import H4MLConverter
+from sklearn.model_selection import train_test_split
+
+LABEL_BKG = 0
+LABEL_PROMPT = 1
+LABEL_NONPROMPT = 2
+
+MAX_BKG_FRAC = 0.4  # max of bkg fraction to keep for training
+
+
+def enforce_list(x):
+    """
+    Helper method to enforce list type
+
+    Parameters
+    -----------------
+    - x: a string or a list of string
+
+    Returns
+    -----------------
+    - x_list if x was not a list (and not None), x itself otherwise
+    """
+
+    if not isinstance(x, list):
+        # handle possible spaces in config file entry
+        x = x.split(",")
+        for i, element in enumerate(x):
+            x[i] = element.strip()
+
+    return x
+
+
+# pylint: disable= too-many-instance-attributes, too-few-public-methods
+class MlTrainer:
+    """
+    Class for ML training and testing
+    """
```

Comment:
Is a class needed? It looks to me that the code can be written as a few functions and a main

Reply:
Actually, when I started writing the code I did using functions only but as I was progressing I felt that it was cleaner to access some information via class attributes rather than by having duplicated function arguments. Hence the class formalism